### PR TITLE
[ADZ-1551] Last Change Check Instant fix

### PR DESCRIPTION
**Pull request by ziemsky**
Adds dedicated node to track metadata used
during import of API Specifications, moving
the website:lastChangeCheckInstant to it.

This is so that the timestamp is not tracked
on the API Specification documents themselves,
as updating it there used to unnecessarily save
the entire documents on each batch job run,
even if the documents themselves did not change.

### CODE REVIEW

File: cms/src/main/java/uk/nhs/digital/apispecs/ApiSpecificationImportMetadata.java [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        private Instant lastChangeCheckInstant;
+        private final String apiSpecJcrId;
+        private boolean specExists;
+
+        private Item(final String apiSpecJcrId, final Instant lastChangeCheckInstant) {
+            this.lastChangeCheckInstant = lastChangeCheckInstant;
+            this.apiSpecJcrId = apiSpecJcrId;
+        }
+
+        public static Item item(final String apiSpecJcrId, final Instant lastChangeCheckInstant) {
```
Comment (harryhmc) [inline cms/src/main/java/uk/nhs/digital/apispecs/ApiSpecificationImportMetadata.java:77]: not sure what benefit this brings over just making the above constructor public?

File: cms/src/main/java/uk/nhs/digital/apispecs/jcr/ApiSpecificationMetadataJcrRepository.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+            wrapCheckedException(session::save);
+        } catch (final Exception e) {
+            throw new RuntimeException("Failed to save API Specification Metadata at " + API_SPEC_METADATA_ROOT_NODE_PATH, e);
+        }
+
+        log.debug("Saving API Specifications metadata: done.");
+    }
+
+    private static ApiSpecificationImportMetadata metadatataFrom(final Node apiSpecMetadataRootNode) {
```
Comment (harryhmc) [inline cms/src/main/java/uk/nhs/digital/apispecs/jcr/ApiSpecificationMetadataJcrRepository.java:63]: typo in "metadatata" :grin:

File: cms/src/main/java/uk/nhs/digital/apispecs/model/SpecificationSyncData.java [hunk 3]
```diff
@@ -60,16 +67,7 @@ public boolean specReportedAsUpdated() {
...
-        // Once the spec is published for the very first time,
-        // its 'published' variant won't have the 'last check time'
-        // recorded yet, but the last publication time is a good
-        // substitute in such instance and helps avoiding a redundant
-        // call to Apigee in that scenario.
-
-        return localSpec.lastChangeCheckInstant()
-            .orElse(localSpec.lastPublicationInstant()
-                .orElse(Instant.EPOCH));
+        return localMetadata().lastChangeCheckInstant();
```
Comment (harryhmc) [inline cms/src/main/java/uk/nhs/digital/apispecs/model/SpecificationSyncData.java:70]: nice that this method can get cleaned up some :smile:

File: cms/src/test/resources/test-data/api-specifications/ApiSpecificationMetadataJcrRepositoryTest/apiSpecMetadataRoot_withActualMetadata.yml [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+# Root API Specifications metadata node under module config node
+# /hippo:configuration/hippo:modules/api-specification-sync/api-specification-sync/hippo:moduleconfig
+
+- path: /hippo:configuration/hippo:modules/api-specification-sync/hippo:moduleconfig/api-specification-metadata
+  primaryType: hipposys:moduleconfig
+
+# Metadata of API Specification A
+- path: /hippo:configuration/hippo:modules/api-specification-sync/hippo:moduleconfig/api-specification-metadata/api-spec-a-6c684d98-61e9-4ff6-bf40-5e288bd692f8
+  jcr:primaryType: hipposys:moduleconfig
```
Comment (harryhmc) [inline cms/src/test/resources/test-data/api-specifications/ApiSpecificationMetadataJcrRepositoryTest/apiSpecMetadataRoot_withActualMetadata.yml:11]: is there any purpose to the `jcr` prefix being dropped in the other test metadata?

File: cms/src/test/java/uk/nhs/digital/apispecs/ApiSpecificationImportMetadataTest.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        // when
+        final ApiSpecificationImportMetadata actualApiSpecificationImportMetadata = metadataWith(expectedItems);
+
+        // then
+        final List<ApiSpecificationImportMetadata.Item> actualItems = new ArrayList<>(actualApiSpecificationImportMetadata.items());
+
+        assertThat(
+            "Metadata created as empty returns empty collection of items.",
+            actualItems,
+            is(emptyList())
```
Comment (harryhmc) [inline cms/src/test/java/uk/nhs/digital/apispecs/ApiSpecificationImportMetadataTest.java:63]: you're not actually using `expectedItems` in your assertion which i assume was your intent

File: cms/src/test/java/uk/nhs/digital/apispecs/swagger/SwaggerCodeGenOpenApiSpecificationJsonToHtmlConverterTest.java [hunk 6]
```diff
@@ -387,17 +387,36 @@ public void rendersOperationIds_normalisedToContiguousString() {
...
         final String actualSpecHtml = swaggerCodeGenApiSpecHtmlProvider.htmlFrom(specificationJson);
 
         // then
+        // @formatter:off
         assertThat(
             "OperationID rendered as it is on the specification - Camel case and snake case tested.",
-            ignoringUuids(ignoringWhiteSpacesIn(actualSpecHtml)),
-            is(ignoringUuids(ignoringWhiteSpacesIn(expectedSpecHtml)))
+            actualSpecHtml,
+            stringContainsInOrder(
```
Comment (harryhmc) [inline cms/src/test/java/uk/nhs/digital/apispecs/swagger/SwaggerCodeGenOpenApiSpecificationJsonToHtmlConverterTest.java:398]: great work making this test less fragile and tightening the assertion logic

**Review by harryhmc — APPROVED** [review]
note my comments, but nothing breaking here and LGTM as a whole

File: cms/src/main/java/uk/nhs/digital/apispecs/ApiSpecificationImportMetadata.java [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+import java.util.*;
+
+/**
+ * Meta data of API Specifications processed during import.
+ */
+public class ApiSpecificationImportMetadata {
+
+    final Map<String, Item> apiSpecsMetadata = new HashMap<>();
+
+    private ApiSpecificationImportMetadata(final List<Item> apiSpecsMetadata) {
```
[the remark this answers is not in the document]
Comment (ziemsky) [reply cms/src/main/java/uk/nhs/digital/apispecs/ApiSpecificationImportMetadata.java:17]: Indeed, I'm struggling with an elegant container/item distinction here.

The best I can come up with for the parameter name is to rename it to `apiSpecMetadataItems`.

File: cms/src/main/java/uk/nhs/digital/apispecs/ApiSpecificationImportMetadata.java (hunk 1 above)
Comment (ziemsky) [reply cms/src/main/java/uk/nhs/digital/apispecs/ApiSpecificationImportMetadata.java:77]: Admittedly, not much in this particular context, although this does help us avoid the noise of using keyword `new` everywhere.

It's just my personal preference, but it feels like this piece of test code reads _sligthly_ better (especially if we were to create a number of items in one go):
```
        final ApiSpecificationImportMetadata newApiSpecificationImportMetadata = metadataWith(
            item(apiSpecJcrId,  Instant.parse(oldLocalCheckTime))
        );
```

In general, however, I feel that factory methods tend to offer more expressiveness and control over object creation in comparison to always using constructor, and that's what has gotten me into habit of using them almost everywhere.

File: cms/src/test/resources/test-data/api-specifications/ApiSpecificationMetadataJcrRepositoryTest/apiSpecMetadataRoot_withActualMetadata.yml (hunk 4 above)
Comment (ziemsky) [reply cms/src/test/resources/test-data/api-specifications/ApiSpecificationMetadataJcrRepositoryTest/apiSpecMetadataRoot_withActualMetadata.yml:11]: Another good one!

Yes, it's actually the entries without the `jcr:` prefix that are more correct, as this is what class uk/nhs/digital/test/util/MockJcrRepoProvider.java:45 expects.

File: cms/src/test/java/uk/nhs/digital/apispecs/ApiSpecificationImportMetadataTest.java (hunk 5 above)
Comment (ziemsky) [reply cms/src/test/java/uk/nhs/digital/apispecs/ApiSpecificationImportMetadataTest.java:63]: Hmm, indeed, I'm not making _direct_ use of them, which makes the naming a bit confusing.

Admittedly, the intent was to expect an empty list. I'll clean it up.

File: cms/src/test/java/uk/nhs/digital/apispecs/swagger/SwaggerCodeGenOpenApiSpecificationJsonToHtmlConverterTest.java (hunk 6 above)
Comment (ziemsky) [reply cms/src/test/java/uk/nhs/digital/apispecs/swagger/SwaggerCodeGenOpenApiSpecificationJsonToHtmlConverterTest.java:398]: Thanks, although this sword has two edges, as I have learned the hard way.

When HTML result is large and the test fails, this approach actually makes it more difficult to spot what exactly the mismatch is. I was one third into streamlining `rendersAllSupportedRequestAndResponseParameters` this way and had to back out when I realised this.

This approach seems fine for smaller payloads but has some cost with larger ones. I may give `rendersAllSupportedRequestAndResponseParameters` another go when I have more time to think about it.

**Review by harryhmc — APPROVED** [review]
LGTM. nice work rafal

**Comment by ziemsky** [comment]
`dafd2d1` is just a rebase onto latest `master`.

**Pull request merged**